mlir/lib/Conversion/AIRRtToNpuPass.cpp

```diff
@@ -137,22 +137,32 @@ struct DmaToNpuPattern : public OpConversionPattern<DmaMemcpyNdOp> {
           .getResult();
     };
     SmallVector<Value> offsets;
-    SmallVector<int64_t> staticOffsets;
-    if (auto const_int = getConstantIntValue(adaptor.getOffset3()))
-      staticOffsets.push_back(*const_int);
-    else
+    SmallVector<int64_t>
+        staticOffsets; // Note: for static offsets we compose one single offset
+                       // at the last dimension.
```

Comment:
What is the motivation for this? Should it be a canonicalization in mlir-aie instead?

Reply:
I noticed an issue in mlir-aie, where the multi-dimensional offsets were composed to the BD's base offset using memref shape: https://github.com/Xilinx/mlir-aie/blob/d850560c77799af96c6361a79e34cb0a8e842c50/lib/Dialect/AIEX/Transforms/AIEDmaToNpu.cpp#L284

I think this would not work when the memref shape and data movement wraps do not match. We have never had this issue before because with matmul they would always match; direct code-generated conv2d starts to trigger this issue.

So to get around it I compose the offsets in mlir-air, into one single static offset at the lowest dimension.

Reply:
Why don't we fix the mlir-aie code to work in the general case instead of adding workarounds?

Reply:
Make sense. Seems related: https://github.com/Xilinx/mlir-aie/issues/1578